runtime/src/accounts_db.rs

```diff
@@ -1381,9 +1416,68 @@ impl AccountsDB {
             .and_then(|account_storage_entry| account_storage_entry.get_account(account_info))
     }
 
-    fn find_storage_candidate(&self, slot: Slot) -> Arc<AccountStorageEntry> {
+    fn try_recycle_and_insert_store(
+        &self,
+        slot: Slot,
+        min_size: u64,
+        max_size: u64,
+    ) -> Option<Arc<AccountStorageEntry>> {
+        let store = self.try_recycle_store(slot, min_size, max_size)?;
+        self.insert_store(slot, store.clone());
+        Some(store)
+    }
+
+    fn try_recycle_store(
+        &self,
+        slot: Slot,
+        min_size: u64,
+        max_size: u64,
+    ) -> Option<Arc<AccountStorageEntry>> {
+        let mut max = 0;
+        let mut min = std::u64::MAX;
+        let mut avail = 0;
+        let mut recycle_stores = self.recycle_stores.write().unwrap();
+        for (i, store) in recycle_stores.iter().enumerate() {
+            if Arc::strong_count(store) == 1 {
+                max = std::cmp::max(store.accounts.capacity(), max);
+                min = std::cmp::min(store.accounts.capacity(), min);
+                avail += 1;
+
+                if store.accounts.capacity() >= min_size && store.accounts.capacity() < max_size {
+                    let old_id = store.append_vec_id();
+                    store.recycle(slot, self.next_id.fetch_add(1, Ordering::Relaxed));
+                    let ret = recycle_stores.swap_remove(i);
```

Comment:
nit: I think we could `swap_remove()`, drop the lock, then `store.recycle()`

Reply:
re-ordered